feat: db pull implementation

svetch opened the pull request:
## Summary  
This PR introduces **`db pull` support to `zenstack_v3`**, similar to Prisma’s `db pull`.  
The command enables introspection of the database schema and generates corresponding models within the project.  

## Current Status  
- [x] Initial implementation of `db pull` command  
- [x] Relation resolving logic
- [x] Implement diff update on db pull
  - [x]  Multi-schema file update(handle schema imports and update model in their schema file)
  - [x] Add/Remove fields on update
  - [x] Add/Remove Models 
  - [x] Update field
  - [x] Update model attributes
- [x] Make all generation option configurable from cli
  - [x] Option to always add `@map` for fields/models
  - [x] Name casing strategy
  - [x] quote, indent 
- [x] improve logging
- [x] Handle the database type mapping per database provider(`@db.\<Raw databse type\>`, ie: `@db.Real`)
- [x] Test on multiple database provider with all zenstack features

## Notes  
- I appreciate any feedback, suggestions, or requests for changes 🙏

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * CLI: Added `db pull` to introspect MySQL/Postgres/SQLite and generate ZenStack models with output, casing, mapping, quoting, and indent options.

* **New API**
  * Fluent factory API for programmatic construction of language ASTs and model generation.

* **Bug Fixes**
  * Safer schema loading with optional import merging and optional access to parsing services.

* **Tests**
  * Extensive provider-aware pull/push tests; many tests converted to async and test utilities improved.

* **Chores**
  * Added DB providers, casing utilities, packaging exports, and codegen/formatting improvements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

svetch committed:
feat: initial db pull implementation

svetch committed:
fix: generate imports and attributes for zmodel-code-generator

svetch committed:
fix: add option to not exclude imports in loadDocument

svetch committed:
fix: continue work on db pull

svetch committed:
fix: missing import

svetch committed:
fix: rewrite model generation

generate model  from ground up and diff later

svetch committed:
feat: add ast factory

svetch committed:
fix: ast factory import order

svetch committed:
fix: some runtime bugs

svetch committed:
fix: lint fix

svetch committed:
fix: update zmodel code generator

- include imports in output
- fix indentaions
- include comments in output

svetch committed:
feat: add exclude schemas option

svetch committed:
feat: implement initial diff update

svetch committed:
fix: update format in zmodel code generator

svetch committed:
fix: typo

svetch commented on packages/cli/src/actions/pull/index.ts:
```diff
@@ -268,21 +255,65 @@ export function syncTable({
-    model.declarations.push(modelFactory.node);
-
-    table.indexes.forEach((index) => {
+    const uniqieColumns = table.columns.filter((c) => c.unique && !c.pk).map((c) => c.name);
```
Fixed, thank you!

svetch committed:
feat: progress on database introspection and syncing

svetch committed:
fix: make ignore behave it does in prisma with no index models

svetch committed:
fix: lint fix

svetch committed:
feat: make all format options configurable

svetch committed:
fix: lint fix

svetch committed:
feat:  Handle the database type mapping

ymc9 commented:
Just a heads up, I'm renaming "@zenstackhq/runtime" package to "@zenstackhq/orm".

svetch committed:
fix: catch up with feature updates

- improve code styling
- enable schema support for db pull

svetch committed:
fix: add sqlite e2e test and fix some bugs

svetch committed:
fix: lint fix

svetch committed:
fix: formatting for e2e test schemas

svetch committed:
test: run db pull e2e test also for postgres

svetch committed:
fix: postgres instorspection schema filter

svetch committed:
test: update cli tests

svetch committed:
feat(cli): Improves database introspection and syncing

Enhances the `db pull` command with a spinner for better UX.
Adds color-coded logging to highlight important steps.
Provides more detailed output on schema changes, including deleted models, enums, added fields, and deleted attributes.
Also includes minor improvements to enum mapping and constraint handling.

svetch committed:
fix(cli): fixes field casing and sort issues

svetch committed:
chore(cli): remove temporary test script

Deletes an unused script used for experimenting with URI path resolution. Cleans up the codebase by removing development-only artifacts.

svetch committed:
chore: update pnpm-lock.yaml

svetch committed:
feat(cli): add MySQL support for schema introspection

Introduces a MySQL-specific introspection provider to support pulling existing database schemas into ZenStack.

The implementation includes logic for mapping MySQL data types to ZenStack types, handling auto-incrementing fields, and parsing MySQL-specific enum definitions. It utilizes dynamic imports for database drivers to minimize the CLI footprint for users not targeting MySQL.

svetch committed:
fix(cli): improve field matching logic during db pull

svetch committed:
feat(cli): enhance SQLite introspection with autoincrement support

svetch committed:
fix(cli): refine attribute generation in db pull

svetch committed:
test(cli): update db pull tests for SQLite specific behavior

svetch committed:
refactor(language): export ZModelServices type

svetch committed:
fix(cli): improve sqlite introspection for autoincrement and fk names

svetch committed:
feat(cli): enhance field matching logic during pull by using relation fields

svetch committed:
refactor(cli): refine relation name generation and table syncing

svetch committed:
test(cli): update pull tests to reflect improved schema generation

svetch committed:
test(cli): add MySQL support to test utility helpers

Extends the testing infrastructure to support MySQL databases.

Adds MySQL configuration defaults and environment variable overrides.
Updates the prelude generation logic to handle MySQL connection strings and provider types, enabling broader database integration testing across the CLI.

svetch committed:
fix(cli): omit default constraint names in table sync

Avoids explicitly declaring unique constraint names when they match the default database naming convention. This results in cleaner generated schema code by removing redundant mapping arguments.

svetch committed:
fix: correctly handle default values for 'text' type in PostgreSQL

svetch committed:
test: fix typo in pull test description

svetch committed:
refactor: dynamically determine supported db providers in CLI

svetch committed:
fix: sort table indexes to ensure stable schema generation

svetch committed:
chore(cli): remove debug artifacts and silence test logs

Removes hardcoded file system path debugging and unnecessary console logging from the introspector and test suites. Silences CLI command output during tests to provide a cleaner test execution environment.

svetch committed:
fix(cli): ensure MySQL column and index ordering

Wraps JSON_ARRAYAGG calls in subqueries with explicit ORDER BY clauses to maintain correct metadata ordering.

This addresses a limitation in MySQL versions prior to 8.0.21, where ORDER BY is not supported directly within the JSON_ARRAYAGG function, ensuring consistent introspection results across different database versions.

svetch committed:
fix(cli): preserve column order during MySQL pull

Ensures database columns are sorted by their ordinal position during the introspection process. This maintains the original schema structure and provides a consistent output that matches the physical database layout.

svetch committed:
refactor(cli): remove schema fields from MySQL queries

Eliminates redundant schema and database name fields from the MySQL introspection query. Since MySQL does not support multi-schema architectures internal to a single connection in this context, removing these fields simplifies the data structure and avoids unnecessary metadata overhead.

svetch committed:
fix(cli): improve MySQL introspection and index mapping

Refines the database pull process to better handle MySQL-specific patterns. Improves unique constraint detection to prevent redundant mapping attributes when default naming conventions are used.

Updates the MySQL introspection logic to correctly identify boolean types, handle timestamp precision in default values, and normalize numeric defaults. Also ensures auto-incrementing columns and primary key indexes are correctly mapped to prevent schema duplication.

svetch committed:
test(cli): pass provider to default prelude in tests

Ensures that the default schema prelude correctly reflects the database provider specified in test options. This prevents inconsistencies when generating test projects with non-default providers.

svetch committed:
fix(cli): improve MySQL introspection for types and defaults

Disables NativeEnum support for MySQL to prevent loss of schema-level enums since MySQL enums are column-specific.

Refines boolean and numeric type mapping to better handle synthetic boolean types and preserve decimal precision in default values.

Updates default value parsing logic to correctly identify unquoted strings and avoid misinterpreting numeric literals as booleans.

svetch committed:
fix(cli): improve MySQL default value introspection

Refines how default values are handled during database introspection for MySQL by considering the specific field type. This ensures that boolean variants and numeric literals for Float and Decimal types are correctly formatted and preserved.

Also clarifies unsupported features in the SQLite provider to improve codebase maintainability.

svetch committed:
test(cli): expand and reorganize db pull tests

Enhances the test suite for the database pull command by adding comprehensive coverage for common schema features and PostgreSQL-specific functionality.

Includes new test cases for:
- Restoring complex schemas from scratch, including relations and indexes
- Preserving existing imports in multi-file schema setups
- Handling PostgreSQL-specific features like multi-schema support and native enums
- Verifying schema preservation for field and table mappings

The tests are restructured for better clarity across different database providers.

svetch committed:
feat: modernize PostgreSQL introspection provider

svetch committed:
feat: modernize MySQL introspection provider

svetch committed:
refactor: restructure introspection provider interface and attribute generation

svetch committed:
feat: track imports and auto-format during db pull

svetch committed:
fix: improve relation field naming and default action handling

svetch committed:
feat: modernize SQLite introspection provider

svetch committed:
test: update pull tests to reflect naming and formatting improvements

svetch committed:
fix(cli): refactor PostgreSQL type casting and fix index order

Extracts PostgreSQL type casting logic into a reusable helper function to improve maintainability and ensure consistent attribute handling across all field types.

Adjusts the table index sorting logic to better preserve the original database creation order while maintaining the priority of unique indexes.

svetch committed:
fix(cli): filter out auto-generated MySQL indexes

Prevents foreign key indexes created automatically by MySQL from appearing in the introspected schema. This ensures the output reflects manually defined indexes and avoids redundancy in schema definitions.

svetch committed:
test(cli): support datasource extras in test utils

Enhances the test utility helpers to allow passing extra datasource properties, such as multi-schema configurations for PostgreSQL.

Refactors existing database pull tests to use these extra properties, ensuring the generated ZModel schema correctly reflects multi-schema environments while simplifying assertions.

svetch committed:
fix: address PR comments

ymc9 commented on packages/language/src/document.ts:
```diff
@@ -121,14 +123,16 @@ export async function loadDocument(
+    }
 
     // extra validation after merging imported declarations
     const additionalErrors = validationAfterImportMerge(model);
```
I'm not sure how complete semantic checks need to be for "non-merging" models. Strictly speaking, we can also extend the validation here to traverse all `imports` (if merging is turned off) and detect conflicts, but I'n not sure if it matters for your use case.

ymc9 commented on packages/cli/src/index.ts:
```diff
@@ -143,6 +143,31 @@ function createProgram() {
+            ),
+        )
+        .addOption(
+            new Option('--field-casing <pascal|camel|snake|kebab|none>', 'set the casing of generated fields').default(
```
I believe kebab casing is not valid for ZModel declarations.

ymc9 commented on packages/cli/src/actions/db.ts:
```diff
@@ -49,3 +78,347 @@ async function runPush(options: Options) {
+                    model.declarations.splice(index, 1);
+                    deletedEnums.push(colors.red(`- Enum ${decl.name} deleted`));
+                });
+        //
```
Missing a comment here?

ymc9 commented on packages/cli/src/actions/db.ts:
```diff
@@ -49,3 +78,347 @@ async function runPush(options: Options) {
+                        // Try matching by type reference
+                        originalFields = originalDataModel.fields.filter(
+                            (d) =>
+                                f.$type === 'DataField' &&
```
Is this search too broad? A model can have multiple relations to the same model, and for this case, they'll all be identified as original fields, right?

svetch replied:
Never mind, I realized why I used this search method and added comments to the code for future reference. Essentially, we cannot determine with 100% certainty the original field of a relation that lacks the @relation attribute. We need to ensure that the initial pull contains at least one of these fields. In the future, we may need logic to detect and verify the required count of these fields between models. I hope this is a rare scenario, so we'll have time to address it later

ymc9 commented on packages/cli/src/actions/db.ts:
```diff
@@ -49,3 +78,347 @@ async function runPush(options: Options) {
+                    const originalField = originalFields.at(0);
+                    Object.freeze(originalField);
+                    if (!originalField) {
+                        addedFields.push(colors.green(`+ Field ${f.name} added to ${originalDataModel.name}`));
```
For future improvements: I think the log will be easier to read if it's organized hierarchically:

Model changes
  - Field changes
     - Attribute changes

ymc9 reviewed (commented):
Hi @svetch , thanks a lot for making such a comprehensive PR. You must have spent quite lot of time on it. I can see the changes are very carefully crafted.

While I'm still half way into reviewing it, just wanted to share my comments so far. I'll try to finish my review tomorrow.

svetch commented on packages/cli/src/actions/db.ts:
```diff
@@ -49,3 +78,347 @@ async function runPush(options: Options) {
+
+            fs.writeFileSync(outPath, zmodelSchema);
+        } else {
+            for (const { uri, parseResult: { value: model } } of docs) {
```
yes exactly

svetch committed:
fix: address PR comments

ymc9 commented on packages/cli/src/actions/pull/provider/postgresql.ts:
```diff
@@ -0,0 +1,475 @@
+                return true;
+            default:
+                return false;
+        }
```
I think this is a valid comment @svetch

svetch commented on packages/cli/src/actions/db.ts:
```diff
@@ -49,3 +78,347 @@ async function runPush(options: Options) {
+        const { enums: allEnums, tables: allTables } = await provider.introspect(datasource.url);
+        spinner.succeed('Database introspected');
+
+        const enums = provider.isSupportedFeature('Schema')
```
This handles the exclude schema filter. Since SQLite/MySQL don't support schemas, I realized it made more sense to move this logic into the provider itself

svetch commented on packages/cli/src/actions/pull/provider/sqlite.ts:
```diff
@@ -0,0 +1,378 @@
+            case 'Float':
+                if (/^-?\d+\.\d+$/.test(val)) {
+                    const numVal = parseFloat(val);
+                    return (ab) => ab.NumberLiteral.setValue(numVal === Math.floor(numVal) ? numVal.toFixed(1) : String(numVal));
```
I'm not sure about SQLite, but MySQL and PostgreSQL require this for proper formatting. I assume Claude included it just in case

svetch committed:
fix: address PR comments

svetch commented on packages/cli/src/actions/db.ts:
```diff
@@ -49,3 +78,472 @@ async function runPush(options: Options) {
+
+        spinner.start('Introspecting database...');
+        const { enums, tables } = await provider.introspect(datasource.url, { schemas: datasource.allSchemas });
+        spinner.succeed('Database introspected');
```
but the getDatasource function throws and error if the url is undefined or missing

svetch committed:
fix: address PR comments

ymc9 commented on packages/cli/src/actions/pull/provider/postgresql.ts:
```diff
@@ -0,0 +1,597 @@
+            const { rows: enums } = await client.query<IntrospectedEnum>(enumIntrospectionQuery);
+
+            // Filter tables and enums to only include those from the selected schemas
+            const filteredTables = tables.filter((t) => options.schemas.includes(t.schema));
```
Will this cause all tables and enums filtered out if the user didn't specify "allSchemas" in the datasource? In this case, I think we should default to including "public" schema.

ymc9 commented on packages/cli/src/actions/pull/provider/postgresql.ts:
```diff
@@ -0,0 +1,597 @@
+GROUP BY schema_name, enum_type
+ORDER BY schema_name, enum_type;`;
+
+const tableIntrospectionQuery = `
```
This introspection SQL is quite complex and not easy to understand. Could you help add some inline comments?

svetch replied:
I used Claude to help write the description as it generated most of the queries. Can you give it a review? By the way, this is largely based on Prisma Studio’s introspection SQL, but I’ve extended it to include the necessary schema info.

ymc9 replied:
Looks good to me. Btw, could you share the original Prisma source location? Just want to use it as a reference. Thanks! @svetch

svetch replied:
I couldn’t find the direct source code for Prisma Studio, but it uses the @prisma/studio-core npm package with a simple HTTP handler. The Studio sends the introspection SQL down to load the UI correctly. However, I did find the original introspection code for Prisma here: https://github.com/prisma/prisma-engines/blob/main/schema-engine/sql-schema-describer/src/postgres.rs

<details>
  <summary>I also quickly reverse-engineered it studio-core using Chrome DevTools since it includes source maps. Here is the content of the introspection logic:</summary>

```typescript
import { type SelectQueryBuilder, sql } from "kysely";
import { jsonArrayFrom } from "kysely/helpers/postgres";

import { type BuilderRequirements, compile, type QueryResult } from "../query";
import { getPostgreSQLBuilder } from "./builder";

const REGULAR_TABLE_KIND = "r";
const VIEW_TABLE_KIND = "v";
const TABLE_KINDS = [REGULAR_TABLE_KIND, VIEW_TABLE_KIND] as const;
type TableKind = (typeof TABLE_KINDS)[number];

interface Database {
  /** column defaults */
  "pg_catalog.pg_attrdef": {
    /** default expression as a node tree, use pg_get_expr(adbin, adrelid) */
    adbin: string;
    /** column number */
    adnum: number;
    /** table identifier */
    adrelid: number;
  };
  /** columns */
  "pg_catalog.pg_attribute": {
    /**
     * generation flag for computed columns:
     * '' (empty) for regular columns,
     * 's' for STORED generated columns (Postgres GENERATED ALWAYS AS).
     */
    attgenerated: "" | "s";
    /**
     * identity column flag:
     * '' (empty) for regular columns,
     * 'a' for ALWAYS (GENERATED ALWAYS AS IDENTITY),
     * 'd' for BY DEFAULT (GENERATED BY DEFAULT AS IDENTITY).
     */
    attidentity: "" | "a" | "d";
    /** was the column dropped. dropped columns might not be 100% removed immediately. */
    attisdropped: boolean;
    /** column name */
    attname: string;
    /** NOT NULL constraint: true if the column has a NOT NULL constraint */
    attnotnull: boolean;
    /** column order number in its table */
    attnum: number;
    /** table/view identifier */
    attrelid: string;
    /** column datatype identifier */
    atttypid: string;
  };
  /** resources, e.g. tables, sequences, etc. */
  "pg_catalog.pg_class": {
    /** resource id */
    oid: number;
    /** resource kind: regular table, view */
    relkind: TableKind;
    /** resource name */
    relname: string;
    /** schema id */
    relnamespace: number;
  };
  /** constraints */
  "pg_catalog.pg_constraint": {
    /** if a foreign key, the referenced columns */
    confkey: number[];
    /** if a foreign key, the referenced table */
    confrelid: number;
    /**
     * an ordered array of columns included in the constraint.
     *
     * the columns are represented by their ordinal number (`attnum`) in their relation (table).
     *
     * @example `'{1,3}'` - a constraint with 2 columns, the 1st & 3rd in their table.
     */
    conkey: number[];
    /** relation (table) id which this constraint was created for */
    conrelid: number;
    /**
     * constraint type.
     *
     * `'p'` - primary key constraint.
     * `'f'` - foreign key constraint.
     */
    contype: "p" | "f" | (string & {});
  };
  /** enum types' values */
  "pg_catalog.pg_enum": {
    /** enum value */
    enumlabel: string;
    /** enum's type id */
    enumtypid: number;
  };
  /** schemas */
  "pg_catalog.pg_namespace": {
    /** schema name */
    nspname: string;
    /** schema id */
    oid: number;
  };
  /** datatypes */
  "pg_catalog.pg_type": {
    /** type id */
    oid: number;
    /** type name */
    typname: string;
    /** type's schema (aka namespace) id */
    typnamespace: number;
  };
}

/**
 * Returns a query that returns metadata for all user-defined tables and views in the database.
 */
export function getTablesQuery(requirements?: Omit<BuilderRequirements, "Adapter" | "QueryCompiler">) {
  return compile(
    getPostgreSQLBuilder<Database>(requirements)
      .selectFrom("pg_catalog.pg_class as cls")
      .innerJoin("pg_catalog.pg_namespace as ns", "cls.relnamespace", "ns.oid")
      .$call(excludeSystemTables)
      // regular tables or views
      .where("cls.relkind", "in", TABLE_KINDS)
      .select((eb) => [
        "ns.nspname as schema",
        "cls.relname as name",
        jsonArrayFrom(
          eb
            .selectFrom("pg_catalog.pg_attribute as att")
            .innerJoin("pg_catalog.pg_type as typ", "typ.oid", "att.atttypid")
            .innerJoin("pg_catalog.pg_namespace as tns", "tns.oid", "typ.typnamespace")
            .leftJoin("pg_catalog.pg_constraint as pk_con", (jb) =>
              jb
                .on("pk_con.contype", "=", "p")
                .onRef("pk_con.conrelid", "=", "cls.oid")
                .on((eb) => eb("att.attnum", "=", eb.fn.any("pk_con.conkey"))),
            )
            .leftJoin("pg_catalog.pg_constraint as fk_con", (jb) =>
              jb
                .on("fk_con.contype", "=", "f")
                .onRef("fk_con.conrelid", "=", "cls.oid")
                .on((eb) => eb("att.attnum", "=", eb.fn.any("fk_con.conkey"))),
            )
            .leftJoin("pg_catalog.pg_class as fk_cls", "fk_cls.oid", "fk_con.confrelid")
            .leftJoin("pg_catalog.pg_namespace as fk_ns", "fk_ns.oid", "fk_cls.relnamespace")
            .leftJoin("pg_catalog.pg_attribute as fk_att", (jb) =>
              jb
                .onRef("fk_att.attrelid", "=", "fk_cls.oid")
                // Use a callback for the .on() condition to ensure correct context for types
                .on((on_eb) => on_eb("fk_att.attnum", "=", on_eb.fn.any("fk_con.confkey"))),
            )
            .leftJoin("pg_catalog.pg_attrdef as def", (jb) =>
              jb.onRef("def.adrelid", "=", "att.attrelid").onRef("def.adnum", "=", "att.attnum"),
            )
            .whereRef("att.attrelid", "=", "cls.oid")
            // exclude system columns
            .where("att.attnum", ">=", 0)
            .where("att.attisdropped", "!=", true)
            .select([
              "att.attname as name",
              "fk_att.attname as fk_column",
              "fk_cls.relname as fk_table",
              "fk_ns.nspname as fk_schema",
              "tns.nspname as datatype_schema",
              "typ.typname as datatype",
            ])
            .select((eb) => [
              eb("att.attidentity", "!=", "")
                .or(
                  eb("def.adbin", "is not", null).and(
                    eb.fn("pg_get_expr", ["def.adbin", "def.adrelid"]),
                    "like",
                    "nextval(%",
                  ),
                )
                .$castTo<boolean>()
                .as("autoinc"),
              eb("att.attgenerated", "!=", "").$castTo<boolean>().as("computed"),
              eb.fn<string | null>("pg_get_expr", ["def.adbin", "def.adrelid"]).as("default"),
              eb("att.attnotnull", "!=", true).$castTo<boolean>().as("nullable"),
              eb.fn
                .coalesce(
                  eb
                    .selectFrom("pg_catalog.pg_enum as enm")
                    .whereRef("enm.enumtypid", "=", "typ.oid")
                    .select((eb) => eb.fn.jsonAgg(eb.ref("enm.enumlabel")).as("o")),
                  sql<string[]>`'[]'`,
                )
                .as("options"),
              eb.fn<number | null>("array_position", ["pk_con.conkey", "att.attnum"]).as("pk"),
            ]),
        ).as("columns"),
      ]),
  );
}

/**
 * For testing purposes.
 */
export function mockTablesQuery() {
  return [
    {
      schema: "zoo",
      name: "animals",
      columns: [
        {
          autoinc: true,
          computed: false,
          datatype: "int4",
          datatype_schema: "pg_catalog",
          default: "nextval('zoo.animals_id_seq'::regclass)",
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "id",
          nullable: false,
          options: [],
          pk: 1,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "text",
          datatype_schema: "pg_catalog",
          default: null,
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "name",
          nullable: true,
          options: [],
          pk: null,
        },
      ],
    },
    {
      schema: "public",
      name: "users",
      columns: [
        {
          autoinc: true,
          computed: false,
          datatype: "int4",
          datatype_schema: "pg_catalog",
          default: "nextval('users_id_seq'::regclass)",
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "id",
          nullable: false,
          options: [],
          pk: 1,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "timestamp",
          datatype_schema: "pg_catalog",
          default: "CURRENT_TIMESTAMP",
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "created_at",
          nullable: true,
          options: [],
          pk: null,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "timestamp",
          datatype_schema: "pg_catalog",
          default: null,
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "deleted_at",
          nullable: true,
          options: [],
          pk: null,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "varchar",
          datatype_schema: "pg_catalog",
          default: null,
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "role",
          nullable: true,
          options: [],
          pk: null,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "varchar",
          datatype_schema: "pg_catalog",
          default: null,
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "name",
          nullable: true,
          options: [],
          pk: null,
        },
        {
          autoinc: false,
          computed: true,
          datatype: "text",
          datatype_schema: "pg_catalog",
          default: null,
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "name_role",
          nullable: false,
          options: [],
          pk: null,
        },
      ],
    },
    {
      schema: "public",
      name: "composite_pk",
      columns: [
        {
          autoinc: false,
          computed: false,
          datatype: "uuid",
          datatype_schema: "pg_catalog",
          default: "gen_random_uuid()",
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "id",
          nullable: false,
          options: [],
          pk: 1,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "text",
          datatype_schema: "pg_catalog",
          default: null,
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "name",
          nullable: true,
          options: [],
          pk: 2,
        },
        {
          autoinc: false,
          computed: false,
          datatype: "timestamp",
          datatype_schema: "pg_catalog",
          default: "now()",
          fk_column: null,
          fk_schema: null,
          fk_table: null,
          name: "created_at",
          nullable: true,
          options: [],
          pk: null,
        },
      ],
    },
  ] as const satisfies QueryResult<typeof getTablesQuery>;
}

/**
 * Returns a query that returns the current timezone setting of the PostgreSQL database.
 */
export function getTimezoneQuery() {
  const builder = getPostgreSQLBuilder();

  return compile(builder.selectNoFrom(builder.fn<string>("current_setting", [sql.lit("timezone")]).as("timezone")));
}

/**
 * For testing purposes.
 */
export function mockTimezoneQuery() {
  return [{ timezone: "UTC" }] as const satisfies QueryResult<typeof getTimezoneQuery>;
}

/**
 * We don't want PostgreSQL internals in introspection results. To be used in `$call`.
 */
function excludeSystemTables<
  T extends SelectQueryBuilder<
    Database & { ns: Database["pg_catalog.pg_namespace"]; cls: Database["pg_catalog.pg_class"] },
    "ns" | "cls",
    object
  >,
>(qb: T): T {
  return qb.where(`ns.nspname`, `!~`, `^pg_`).where(`ns.nspname`, `!=`, `information_schema`) as never;
}

```
</details>

ymc9 commented on packages/cli/src/actions/pull/provider/mysql.ts:
```diff
@@ -0,0 +1,519 @@
+        }
+
+        // Handle enum defaults
+        if (datatype === 'enum' && datatype_name) {
```
There are three branches dealing with enum types in this function - one here, one inside the switch/case, and one after that. Are they all needed? Or is enum default values already fully handled here?

svetch replied:
Fixed

ymc9 commented on packages/cli/src/actions/pull/index.ts:
```diff
@@ -0,0 +1,571 @@
+            model.declarations.push(factory.get({ $container: model }));
+        }
+    } else {
+        oldModel.declarations
```
It seems this part of logic is to clone the original enum AST nodes. If so, Langium's `AstUtils` export has a `copyAstNode` function that does this job. Since we don't care about reference resolution here, I believe you can pass a dummy `buildReference` callback.

svetch replied:
Thank you for the suggestion! It's much cleaner now.

ymc9 commented on packages/cli/test/utils.ts:
```diff
@@ -1,22 +1,104 @@
     return workDir;
 }
 
+export async function createFormattedProject(
```
Shall we unify `createProject` and `createFormattedProject`? I think it's good to always format the zmodel file.

svetch replied:
Totally agree. I was actually thinking the same thing, just wasn't sure if you'd prefer.

ymc9 reviewed (commented):
Hi @svetch , sorry for the delay. I've finished reviewing all remaining files. Please check my comments and questions when you have time. THANKS!

svetch committed:
fix(cli): improve file path resolution in pull action

svetch committed:
refactor(cli): extract and enhance name casing logic

svetch committed:
refactor(cli): consolidate default value normalization

svetch committed:
feat(cli): improve enum syncing and relation naming during pull

svetch committed:
docs(cli): add documentation comments to SQL introspection queries

svetch committed:
test(cli): refactor test utilities and modernize test suites

svetch committed:
fix(cli): improve db pull for composite FKs and MySQL uniqueness

Enhances database introspection to correctly handle composite foreign keys by mapping columns by position rather than name alone.

Improves MySQL introspection by checking statistics tables for single-column unique indexes, ensuring accurate model generation even when column keys are ambiguous.

Ensures MySQL synthetic enum names respect requested model casing to prevent unnecessary schema mapping.

Adds comprehensive tests for composite relations and database-specific uniqueness detection.

svetch commented on packages/cli/test/db/pull.test.ts:
```diff
@@ -0,0 +1,769 @@
+    });
+});
+
+describe('DB pull - SQL specific features', () => {
+    it('should restore enum fields from zero', async ({ skip }) => {
+        const provider = getTestDbProvider();
+        if (provider !== 'mysql' && provider !== 'postgresql') {
+            skip();
+            return;
+        }
+
+        const { workDir, schema } = await createProject(
+            `model User {
+    id     Int    @id @default(autoincrement())
+    email  String @unique
+    status UserStatus @default(ACTIVE)
+}
+
+enum UserStatus {
+    ACTIVE
+    INACTIVE
+    SUSPENDED
+}`);
+        runCli('db push', workDir);
+
+        const schemaFile = path.join(workDir, 'zenstack/schema.zmodel');
+
+        // Remove schema content to simulate restoration from zero
+        fs.writeFileSync(schemaFile, getDefaultPrelude());
+
+        // Pull should fully restore the schema including enum fields
+        runCli('db pull --indent 4', workDir);
+
+        const restoredSchema = getSchema(workDir);
+        expect(restoredSchema).toEqual(schema);
+    });
```
Under the hood, these functions automatically use the current test database provider because it is loaded from environment variables. Providing it manually is unnecessary; the parameter is simply an option to override the provider

svetch commented on packages/cli/src/actions/pull/provider/sqlite.ts:
```diff
@@ -0,0 +1,390 @@
+                }
+                return (ab) => ab.NumberLiteral.setValue(val);
+
+            case 'Float':
```
Fixed

svetch commented on packages/cli/src/actions/pull/provider/postgresql.ts:
```diff
@@ -0,0 +1,597 @@
+};
+
+/**
+ * Maps PostgreSQL typnames (from pg_type.typname) to Prisma native type attribute names.
```
Fixed

svetch commented on packages/cli/src/actions/pull/provider/mysql.ts:
```diff
@@ -0,0 +1,519 @@
+                    .map((col: { options?: string | string[] | null }) => ({
+                        ...col,
+                        // Parse enum options from COLUMN_TYPE if present (e.g., "enum('val1','val2')")
+                        options:
```
Fixed

svetch commented on packages/cli/src/actions/pull/index.ts:
```diff
@@ -0,0 +1,571 @@
+
+    const relationName = `${relation.table}${similarRelations > 0 ? `_${relation.column}` : ''}To${relation.references.table}`;
+
+    const sourceNameFromReference = sourceField.name.toLowerCase().endsWith('id') ? `${resolveNameCasing("camel", sourceField.name.slice(0, -2)).name}${relation.type === 'many'? 's' : ''}` : undefined;
```
Fixed

svetch commented on packages/cli/src/actions/pull/index.ts:
```diff
@@ -0,0 +1,571 @@
+    const tableMapAttribute = getAttributeRef('@@map', services);
+    const modelindexAttribute = getAttributeRef('@@index', services);
+
+    if (
```
Fixed

svetch commented on packages/cli/src/index.ts:
```diff
@@ -143,6 +143,36 @@ function createProgram() {
+            ),
+        )
+        .addOption(
+            new Option('--always-map', 'always add @map and @@map attributes to models and fields').default(false),
```
Our database has a massive schema with around 150 tables, but it lacks a consistent naming convention. By always generating @Map and @@Map, we can immediately rename models and fields in ZenStack using 'Rename Symbol' (F2) without triggering immediate database migrations. This allows us to move faster with a cleaner codebase and perform the actual database schema changes later in a single batch.

It’s a specific use case, so if you feel it’s too niche, I’m open to removing it—especially since AI tools can handle some of this now, though this method is more reliable for bulk changes.

ymc9 replied:
Got it. I don't think it harms, although for a niche use case. Let's keep it as is.

svetch commented on packages/cli/src/actions/pull/index.ts:
```diff
@@ -0,0 +1,571 @@
+        if (index.columns.find((c) => c.expression)) {
+            console.warn(
+                colors.yellow(
+                    `These constraints are not supported by Zenstack. Read more: https://pris.ly/d/check-constraints\n- Model: "${table.name}", constraint: "${index.name}"`,
```
I agree.

svetch committed:
fix: address PR comments

svetch committed:
fix(cli): improve SQLite introspection for untyped columns and composite FKs

Ensures columns with no declared type are correctly mapped to Bytes following SQLite affinity rules, preventing them from being marked as Unsupported.

Updates the DDL parser to correctly identify and map constraint names for composite foreign keys. This ensures that multi-column relations are properly restored during the pull process.

Adds regression tests for both untyped columns and composite foreign key restoration.

svetch commented on packages/cli/src/actions/pull/provider/sqlite.ts:
```diff
@@ -0,0 +1,392 @@
+                        foreign_key_on_delete: (fk.on_delete as any) ?? null,
+                    });
+                }
+
+                const columns: IntrospectedTable['columns'] = [];
+                for (const c of columnsInfo) {
+                    // hidden: 1 (hidden/internal) -> skip; 2 (generated) -> mark computed
+                    const hidden = c.hidden ?? 0;
+                    if (hidden === 1) continue;
+
+                    const fk = fkByColumn.get(c.name);
+
+                    // Determine default value - check for autoincrement
+                    // AUTOINCREMENT in SQLite can only be on INTEGER PRIMARY KEY column
+                    let defaultValue = c.dflt_value;
+                    if (hasAutoIncrement && c.pk) {
+                        defaultValue = 'autoincrement';
+                    }
+
+                    columns.push({
+                        name: c.name,
+                        datatype: c.type || '',
+                        datatype_name: null, // SQLite doesn't support native enums
+                        length: null,
+                        precision: null,
+                        datatype_schema: schema,
+                        foreign_key_schema: fk?.foreign_key_schema ?? null,
+                        foreign_key_table: fk?.foreign_key_table ?? null,
+                        foreign_key_column: fk?.foreign_key_column ?? null,
+                        foreign_key_name: fk?.foreign_key_name ?? null,
+                        foreign_key_on_update: fk?.foreign_key_on_update ?? null,
+                        foreign_key_on_delete: fk?.foreign_key_on_delete ?? null,
+                        pk: !!c.pk,
+                        computed: hidden === 2,
+                        nullable: c.notnull !== 1,
+                        default: defaultValue,
+                        unique: uniqueSingleColumn.has(c.name),
+                        unique_name: null,
+                    });
+                }
```
can you share a reproduction with steps?

svetch replied:
I have added support for computed fields across all providers; however, I’ve encountered some design challenges.

Currently, we need to set the type as `Unsupported("varchar(511) GENERATED ALWAYS AS ...")` to ensure `prisma migrate` handles these types correctly. 

@ymc9, should we implement a custom solution to convert these into the "unsupported" format for migration support? I’m unsure of the best approach—perhaps a custom attribute like `@dbComputed("...")` or something similar?

For example:
```prisma
fullName String @dbComputed("GENERATED ALWAYS AS...")
```
Note: Prisma currently excludes these fields during a `db pull`. (tested on sqlite)

ymc9 replied:
Hi @svetch , I haven't really thought much about virtual and generated columns introspection yet. Just to keep the initial implementation from becoming too complex, maybe we can skip them for now and iterate in future releases? What do you think?

svetch replied:
Yeah, I agree with that. For now, I'll keep the basic implementation (introspect as unsupported type) and then we can revisit this in the future. Thank you!

svetch committed:
feat(cli): pull generated/computed columns as Unsupported type

Improves database introspection by identifying generated columns in MySQL, PostgreSQL, and SQLite. These columns are now pulled as `Unsupported` types containing their full DDL definition, preventing issues where read-only database fields were incorrectly treated as writable application-level fields.

Includes normalization for expression formatting and a fix for string literal escaping in the code generator to ensure stable schema output.

Relates to ZModel introspection consistency.

svetch commented on packages/cli/src/actions/pull/provider/postgresql.ts:
```diff
@@ -0,0 +1,653 @@
+        const val = defaultValue.trim();
+
+        // Handle enum defaults (PostgreSQL returns 'value'::enum_type)
+        if (datatype === 'enum' && datatype_name) {
+            const enumDef = enums.find((e) => getDbName(e) === datatype_name);
+            if (enumDef) {
+                // Extract the enum value from the default (format: 'VALUE'::"enum_type")
+                const enumValue = val.replace(/'/g, '').split('::')[0]?.trim();
```
I don't think we should support that. Using apostrophes in enum values doesn't seem like a common or practical use case.

ymc9 commented on packages/cli/src/actions/pull/provider/mysql.ts:
```diff
@@ -0,0 +1,597 @@
+    },
+};
+
+function getTableIntrospectionQuery(databaseName: string) {
```
The two SQL snippets for MySQL use string interpolation. Shall we change them to parameterized queries for peace of mind?

svetch replied:
Fixed

ymc9 reviewed (changes requested):
Adding some final review comments 😄

svetch committed:
fix(cli): Use parameterized queries for MySQL introspection

Switches from template literal interpolation to parameterized queries in MySQL introspection functions. This improves security by preventing potential SQL injection and ensures better handling of database names containing special characters.

svetch committed:
fix(cli): use nullish coalescing for precision check

svetch commented on packages/cli/src/actions/pull/provider/postgresql.ts:
```diff
@@ -0,0 +1,653 @@
+  AND "ns"."nspname" != 'information_schema'  -- exclude the information_schema
+  AND "cls"."relkind" IN ('r', 'v')           -- only tables ('r') and views ('v')
+  AND "cls"."relname" !~ '^pg_'               -- exclude system tables starting with pg_
+  AND "cls"."relname" !~ '_prisma_migrations' -- exclude Prisma migration tracking table
```
I don't think this is a real problem. No one wants to see backup tables in their schema. If they do, we can always tweak this in the future.

ymc9 reviewed (approved):
Thanks for making this amazing contribution @svetch ! I'm merging it and will make it part of v3.4.0 as an experimental feature.
